### PR TITLE
Add member replace e2e test

### PR DESCRIPTION
**Pull request by ZhouJianMS**
This PR address #17069

### CODE REVIEW

**Comment by serathius** [comment]
`grpcProxy-tests / test` failed. Looks like this test is incompatible with grpc-proxy. Recommend moving the test to separate file `ctl_v3_member_no_proxy_test.go` and adding  `//go:build !cluster_proxy` build tag at the top.

**Comment by serathius** [comment]
`Static Analysis / run` failed. Please run `make fix-lint`

File: tests/e2e/ctl_v3_member_test.go [hunk 1]
```diff
@@ -309,3 +312,104 @@ func TestRemoveNonExistingMember(t *testing.T) {
...
+	time.Sleep(etcdserver.HealthInterval)
+	_, err = c.MemberRemove(ctx, memberID)
+	assert.NoError(t, err)
+	_, found, err = getID(ctx, c, member.Config().Name)
+	assert.NoError(t, err)
+	assert.Equal(t, found, false, "Expected member to be removed")
+
+	t.Logf("Removing member %s", member.Config().Name)
+	for member.IsRunning() {
+		member.Kill()
```
Comment (serathius) [inline tests/e2e/ctl_v3_member_test.go:348]: Double checked the Kill here is not needed (mistake in the original test), the removed member should exit by themselves. You can just replace the whole block with `member.Wait(ctx)` to confirm that member exits.

File: tests/e2e/ctl_v3_member_test.go [hunk 2]
```diff
@@ -309,3 +312,104 @@ func TestRemoveNonExistingMember(t *testing.T) {
 	err = epc.Restart(ctx)
 	assert.NoError(t, err)
 }
+
+func TestMemberReplace(t *testing.T) {
+    e2e.BeforeTest(t)
+    ctx := context.Background()
```
Comment (serathius) [inline tests/e2e/ctl_v3_member_test.go:318]: Please provide some timeout for the context.

File: tests/e2e/ctl_v3_member_test.go [hunk 3]
```diff
@@ -309,3 +312,104 @@ func TestRemoveNonExistingMember(t *testing.T) {
...
+			t.Fail()
+		default:
+		}
+		reqCtx, cancel := context.WithTimeout(ctx, time.Second)
+		_, err = cc.MemberAdd(reqCtx, member.Config().Name, []string{removedMemberPeerUrl})
+		cancel()
+		if err == nil {
+			break
+		}
+	}
```
Comment (serathius) [inline tests/e2e/ctl_v3_member_test.go:372]: No need to retry here. Compared to robustness tests there is no other load that could cause request to fail.

File: tests/e2e/ctl_v3_member_test.go [hunk 4]
```diff
@@ -309,3 +312,104 @@ func TestRemoveNonExistingMember(t *testing.T) {
...
+	t.Logf("Adding member %s back", member.Config().Name)
+	removedMemberPeerUrl := member.Config().PeerURL.String()
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fail()
+		default:
+		}
+		reqCtx, cancel := context.WithTimeout(ctx, time.Second)
+		_, err = cc.MemberAdd(reqCtx, member.Config().Name, []string{removedMemberPeerUrl})
```
Comment (serathius) [inline tests/e2e/ctl_v3_member_test.go:367]: Very interesting. I have used this test to validate https://github.com/etcd-io/etcd/pull/17056, looks like the issue is triggered by the immidiate `MemberAdd` request. When I removed the for loop and added a `time.Sleep(time.Second)` issue no longer reproduces.

This makes me think we have risk of potential race condition on membership changes that was is not covered by test at all. cc @ahrtr

**Review by serathius — APPROVED** [review]

File: tests/e2e/ctl_v3_member_no_proxy_test.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+		default:
+		}
+		_, found, err := getID(ctx, c, memberName)
+		if err != nil {
+			continue
+		}
+		if found {
+			break
+		}
+	}
```
Comment (ahrtr) [inline tests/e2e/ctl_v3_member_no_proxy_test.go:100]:
```suggestion
	testutils.ExecuteUntil(ctx, t, func() {
	    for {
	         _, found, err := getMemberIdByName(ctx, c, memberName)
	         if err != nil || !found {
	             time.Sleep(10*time.Millisecond)
	             continue
	         }
	         break
	    }
	})
```

**Comment by ahrtr** [comment]
The test is correct by itself. But I don't see how it can prevent https://github.com/etcd-io/etcd/pull/17017 from happening again.

Proposed actions:
- Merge this PR (after the comments are resolved);
- Create an e2e or integration test (or reuse any existing test) on top of https://github.com/etcd-io/etcd/pull/17017 to reproduce the issue https://github.com/etcd-io/etcd/issues/17052#issuecomment-1837518872.  Afterwards, I am happy to work with anyone to have a deep dive to figure out the root cause.

File: tests/e2e/ctl_v3_member_test.go (hunk 4 above)
Comment (ZhouJianMS) [reply tests/e2e/ctl_v3_member_test.go]: Initially, I was using `assert.NoError()`, which allowed the test to continue executing even when errors occurred in the `MemberAdd` request. Upon logging the error message, I found that

>     ctl_v3_member_no_proxy_test.go:92: Adding member TestMemberReplace-test-2 back failed: match not found.  Set EXPECT_DEBUG for more info Errs: [unexpected exit code [1] after running [/home/zhoujian/etcd-test/etcd/bin/etcdctl --endpoints=http://localhost:20000,http://localhost:20005 member add TestMemberReplace-test-2 --peer-urls http://localhost:20011 -w json]], last lines:
>         {"level":"warn","ts":"2023-12-12T14:35:14.017056+0800","logger":"etcd-client","caller":"v3@v3.6.0-alpha.0/retry_interceptor.go:65","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001c2780/localhost:20000","method":"/etcdserverpb.Cluster/MemberAdd","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: Peer URLs already exists"}
>         Error: etcdserver: Peer URLs already exists

In your case, a single `MemberAdd` request might fail, yet the test continues. Consequently, the cluster doesn't re-add the member, and it loops to locate the corresponding member ID until a timeout occurs. This divergence could arise from issue #17017, which might not get the updated `membersMap` timely to check peer urls.

**Comment by ZhouJianMS** [comment]
> The test is correct by itself. But I don't see how it can prevent #17017 from happening again.
> 
> Proposed actions:
> 
> * Merge this PR (after the comments are resolved);
> * Create an e2e or integration test (or reuse any existing test) on top of [Switch to validating v3 when v2 and v3 are synchronized #17017](https://github.com/etcd-io/etcd/pull/17017) to reproduce the issue [Flaky robustness test #17052 (comment)](https://github.com/etcd-io/etcd/issues/17052#issuecomment-1837518872).  Afterwards, I am happy to work with anyone to have a deep dive to figure out the root cause.

Based on #17017, I have reproduced the issue with same test case. You can find it in draft PR #17100. But this kind of reproduce is quite unstable, over 10 times may only have 1 repro. But if removing loop for `MemberAdd` request as this PR does, the repro rate is quite high.  As I discussed here https://github.com/etcd-io/etcd/pull/17079#discussion_r1423631612, there may be some delay before `membersMap` get latest value from method `MustReadMembersFromBackend`. And it can lead to "failed to update; member unknown" as https://github.com/etcd-io/etcd/issues/17052#issuecomment-1837518872 said. 

This PR can prevent get the out-of-date value of `membersMap` for verification, thus preventing similar issue from happening again. If you have further ideas, I am happy to dig deeper for the root cause.

**Review by ahrtr — APPROVED** [review]
/lgtm

**Comment by ahrtr** [comment]
> You can find it in draft PR #17100. But this kind of reproduce is quite unstable, over 10 times may only have 1 repro.

Thanks for the draft PR to reproduce the issue.  Could you rebase that PR? I will take a look later.

**Comment by ahrtr** [comment]
> > You can find it in draft PR #17100. But this kind of reproduce is quite unstable, over 10 times may only have 1 repro.
> 
> Thanks for the draft PR to reproduce the issue. Could you rebase that PR? I will take a look later.

Please ping me after you finish rebasing that PR and can still reproduce the issue. Thanks.

**Comment by ZhouJianMS** [comment]
> > > You can find it in draft PR #17100. But this kind of reproduce is quite unstable, over 10 times may only have 1 repro.
> > 
> > 
> > Thanks for the draft PR to reproduce the issue. Could you rebase that PR? I will take a look later.
> 
> Please ping me after you finish rebasing that PR and can still reproduce the issue. Thanks.

@ahrtr The branch `ZhouJianMS:repro17052` of PR #17100 is already based on the issue #17052. You can switch to that branch and run test `TestMemberReplaceMultiple` to reproduce the issue.

**Comment by ahrtr** [comment]
> @ahrtr The branch `ZhouJianMS:repro17052` of PR #17100 is already based on the issue #17052. You can switch to that branch and run test `TestMemberReplaceMultiple` to reproduce the issue.

Thanks

**Comment by ahrtr** [comment]
@ZhouJianMS could you backport the test cases to 3.5 and. 3.4 as well? I believe 3.5 and 3.4 have similar issue. I think we need to backport https://github.com/etcd-io/etcd/pull/17119 to 3.5 and 3.4 as well.

**Comment by ZhouJianMS** [comment]
> @ZhouJianMS could you backport the test cases to 3.5 and. 3.4 as well? I believe 3.5 and 3.4 have similar issue. I think we need to backport #17119 to 3.5 and 3.4 as well.

Sure. Willing to do.

**Pull request merged**